ipn/localapi/localapi_test.go

```diff
@@ -212,3 +220,95 @@ func TestShouldDenyServeConfigForGOOSAndUserContext(t *testing.T) {
 		})
 	}
 }
+
+func TestServeWatchIPNBus(t *testing.T) {
+	tstest.Replace(t, &validLocalHostForTesting, true)
+
+	tests := []struct {
+		desc                    string
+		permitRead, permitWrite bool
+		mask                    ipn.NotifyWatchOpt
+		wantStatus              int
+	}{
+		{
+			desc:        "no permission",
+			permitRead:  false,
+			permitWrite: false,
+			mask:        ipn.NotifyInitialState,
+			wantStatus:  http.StatusForbidden,
+		},
+		{
+			desc:        "read initial state",
+			permitRead:  true,
+			permitWrite: false,
+			mask:        ipn.NotifyInitialState,
+			wantStatus:  http.StatusForbidden,
+		},
+		{
+			desc:        "read initial state, no private keys",
+			permitRead:  true,
+			permitWrite: false,
+			mask:        ipn.NotifyInitialState | ipn.NotifyNoPrivateKeys,
+			wantStatus:  http.StatusOK,
+		},
+		{
+			desc:        "read initial state, with private keys",
+			permitRead:  true,
+			permitWrite: true,
+			mask:        ipn.NotifyInitialState,
+			wantStatus:  http.StatusOK,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			h := &Handler{
+				PermitRead:  tt.permitRead,
+				PermitWrite: tt.permitWrite,
+				b:           newTestLocalBackend(t),
+			}
+			s := httptest.NewServer(h)
+			defer s.Close()
+			c := s.Client()
+
+			ctx, cancel := context.WithCancel(context.Background())
+			req, err := http.NewRequestWithContext(ctx, "GET", fmt.Sprintf("%s/localapi/v0/watch-ipn-bus?mask=%d", s.URL, tt.mask), nil)
+			if err != nil {
+				t.Fatal(err)
+			}
+			res, err := c.Do(req)
+			if err != nil {
+				t.Fatal(err)
+			}
+			// Cancel the context so that localapi stops streaming IPN bus
+			// updates.
+			cancel()
+			defer res.Body.Close()
```

Comment:
I'd probably move this up 3 lines, right after the `}` for people pattern matching looking for the defer close after the Do.